src/mpi/coll/allgather/allgather.c

```diff
@@ -170,57 +170,83 @@ int MPIR_Allgather_impl(const void *sendbuf, int sendcount, MPI_Datatype sendtyp
 
     if (comm_ptr->comm_kind == MPIR_COMM_KIND__INTRACOMM) {
         /* intracommunicator */
-        switch (MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM) {
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_brucks:
-                mpi_errno =
-                    MPIR_Allgather_intra_brucks(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                                recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_recursive_doubling:
-                mpi_errno =
-                    MPIR_Allgather_intra_recursive_doubling(sendbuf, sendcount, sendtype, recvbuf,
-                                                            recvcount, recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_ring:
-                mpi_errno =
-                    MPIR_Allgather_intra_ring(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                              recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_nb:
-                mpi_errno =
-                    MPIR_Allgather_allcomm_nb(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                              recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_auto:
-                MPL_FALLTHROUGH;
-            default:
-                mpi_errno =
-                    MPIR_Allgather_intra_auto(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                              recvtype, comm_ptr, errflag);
-                break;
+        int i;
+        for (i = 0; MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM[i] != MPIR_CVAR_ENUM_NULL; i++) {
+            switch (MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM[i]) {
```

Comment:
So for this algorithm -- `ALLGATHER_INTRA` -- in fact we are simply using `MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM[0]` and ignoring the rest, right? If the code does that, and I would suggest write the code like that -- i.e. remove the `for-loop` and simply switch on `ALGORITHM[0]`. For the algorithms that we actually use fall-backs, then of course use the `for-loop`. 

Having the `for-loop` but in fact without `for-loop` semantics are confusing.

Using `enumlist` for the CVAR type is fine. I am only concerned about code should read like what it actually does.

Reply:
It's done this way for consistency between all collective operations (see my discussion with Wesley above).  Using `[0]` seems more confusing than using `[i]`.  We should either retain the full array code as above or move back to `enum` instead of `enumlist` for collectives that do not have algorithms that do not work in certain cases.

Reply:
I think I am leaning toward the latter -- use `enum` instead of `enumlist`, and only select `enumlist` where its necessary. You decide.

Reply:
I prefer @pavanbalaji 's solution and to be consistent everywhere. Otherwise we're just adding more complexity later when someone adds an algorithm and forgets to change this code.

Reply:
I prefer the consistent solution too.  It's not so much for internal developers, but as Hui said, it's for users -- as a user if I have to remember which ones takes lists and which ones doesn't, it's going to be messy.